# AIPI 531 - HW 1 Q3

### Elisa Chen
NetID: eyc11

In [2]:
import stable_baselines3
import gym
import numpy as np
import os
from stable_baselines3 import PPO, A2C #policy gradient
from stable_baselines3.common.evaluation import evaluate_policy
from stable_baselines3.common.monitor import Monitor
from stable_baselines3.common.vec_env import VecVideoRecorder, DummyVecEnv
from stable_baselines3.common.env_util import make_vec_env
from stable_baselines3.common.results_plotter import load_results, ts2xy
from stable_baselines3.common.callbacks import BaseCallback, EvalCallback
import base64
from pathlib import Path
from IPython import display as ipythondisplay

## Creating a training and testing gym environment

In [3]:
#HYPERPARAMETERS
game = "CartPole-v1"
training_steps = 10000

In [8]:
train_env = gym.make(game)
test_env = gym.make(game)
test_env = Monitor(test_env, filename=None, allow_early_resets=True)

#initializing a vanilla model
vanilla_model = A2C('MlpPolicy', train_env, verbose = 0, tensorboard_log="./a2c_cartpole_tensorboard/", gae_lambda = 0) 

#initializing a GAE advantage model
gae_model = A2C('MlpPolicy', train_env, verbose = 0, tensorboard_log="./a2c_cartpole_tensorboard/", gae_lambda = 0.85)

#initializing n-step advantage model
nstep_model = A2C('MlpPolicy', train_env, verbose = 0, tensorboard_log="./a2c_cartpole_tensorboard/", gae_lambda = 1, n_steps = 5)

## Evaluating Rewards for Vanilla Model

In [9]:
## Training the agent for 100000 steps
vanilla_model.learn(total_timesteps=training_steps)

In [10]:
## Evaluating the agent using 100 episodes
reward_mean_vanilla, reward_std_vanilla = evaluate_policy(vanilla_model, test_env, n_eval_episodes=100)
print(f'Mean Reward: {np.round(reward_mean_vanilla,2)} +/- {np.round(reward_std_vanilla,2)}')

Mean Reward: 42.05 +/- 15.45


## Evaluating Rewards for N-Step Advantage Model

In [11]:
# Training the agent for 100000 steps
nstep_model.learn(total_timesteps=training_steps)

In [12]:
## Evaluating the agent using 100 episodes
reward_mean_n_step, reward_std_n_step = evaluate_policy(nstep_model, test_env, n_eval_episodes=100)
print(f'Mean Reward: {np.round(reward_mean_n_step,2)} +/- {np.round(reward_std_n_step,2)}')

Mean Reward: 258.18 +/- 29.89


## Evaluating Rewards for GAE Model

In [13]:
# Training the agent for 100000 steps
gae_model.learn(total_timesteps=training_steps)

In [14]:
## Evaluating the agent using 100 episodes
reward_mean_gae, reward_std_gae = evaluate_policy(nstep_model, test_env, n_eval_episodes=100)
print(f'Mean Reward: {np.round(reward_mean_gae,2)} +/- {np.round(reward_std_gae,2)}')

Mean Reward: 251.43 +/- 27.91


## Final Results

In [15]:
print(f"The Mean Reward For The Vanilla Model Is: {np.round(reward_mean_vanilla,2)} +/- {np.round(reward_std_vanilla,2)}")
print(f"The Mean Reward For The N-Step Advantage Model Is: {np.round(reward_mean_n_step,2)} +/- {np.round(reward_std_n_step,2)}")
print(f"The Mean Reward For The GAE Model Is: {np.round(reward_mean_gae,2)} +/- {np.round(reward_std_gae,2)}")

The Mean Reward For The Vanilla Model Is: 42.05 +/- 15.45
The Mean Reward For The N-Step Advantage Model Is: 258.18 +/- 29.89
The Mean Reward For The GAE Model Is: 251.43 +/- 27.91


It appears that the N-Step Advantage Model fits the training data the best with a mean reward of 258.18. The GAE model produces similar results with a mean reward of 251.43. Both N-Step and GAE models have a large standard deviation of ~30 indicating that there might not be a significant difference between the performance of GAE and N-step Advantage models. We can observe that the Vanilla model produces the worst performance with a mean reward of ~42.05.